auto-election for single-node groups

tbg opened the pull request:
whenever MultiRaft creates a new group, and that group has
only the local node as its member, automatically call elections.

added a helper to the to wait for an election event.

adjusted all tests; had to make minor adjustments in some cases,
and one test is disabled since it relied on the absence of HardState
updates. The minor adjustments are usually waiting out the leader
election event, as otherwise tests could wind up flaky, as in #702,
or due to outstanding work on the mechanism of re-proposing Raft
commands (currently, Raft does that, which means commands may by-
pass the response cache and cause test failures).

the KV test now runs in <5sec on my machine; that's a 100% speedup.

upped the tick interval during tests to 100ms, so in reality we do
not seem to rely on ticks anywhere at the moment.
Of course, that also suggests that we'll need to come up with quite
a bunch of new tests to thoroughly test leader leases, which are due
to land soon.

closes #682

tbg committed:
auto-election for single-node groups

whenever MultiRaft creates a new group, and that group has
only the local node as its member, automatically call elections.

added a helper to the to wait for an election event.

adjusted all tests; had to make minor adjustments in some cases,
and one test is disabled since it relied on the absence of HardState
updates. The minor adjustments are usually waiting out the leader
election event, as otherwise tests could wind up flaky, as in #702,
or due to outstanding work on the mechanism of re-proposing Raft
commands (currently, Raft does that, which means commands may by-
pass the response cache and cause test failures).

the KV test now runs in <5sec on my machine; that's a 100% speedup.

upped the tick interval during tests to 100ms, so in reality we do
not seem to rely on ticks anywhere at the moment.
Of course, that also suggests that we'll need to come up with quite
a bunch of new tests to thoroughly test leader leases, which are due
to land soon.

tschottdorf commented:
@bdarnell the double-increment is now back:
```
Waiting for complete gossip network of 9 peerings: 6 7 8 8 9
All nodes verified in the cluster:
node-1 node-2 node-4
```

But that is expected, correct?

tbg committed:
lock in addition to atomic update in util.Stopper

From https://circleci.com/gh/cockroachdb/cockroach/1466.
The issue is that checking atomically for a value and then acting
on it gives no guarantee that the condition still holds.

==================
WARNING: DATA RACE
Write by goroutine 27:
  sync.raceWrite()
      /usr/src/go/src/sync/race.go:41 +0x36
  sync.(*WaitGroup).Wait()
      /usr/src/go/src/sync/waitgroup.go:125 +0x177
  github.com/cockroachdb/cockroach/util.(*Stopper).Stop()
      /go/src/github.com/cockroachdb/cockroach/util/stopper.go:117 +0x6b
  github.com/cockroachdb/cockroach/storage.(*testContext).Stop()
      /go/src/github.com/cockroachdb/cockroach/storage/range_test.go:185 +0x4c
  github.com/cockroachdb/cockroach/storage.TestGCQueueShouldQueue()
      /go/src/github.com/cockroachdb/cockroach/storage/gc_queue_test.go:115 +0x1a07
  testing.tRunner()
      /usr/src/go/src/testing/testing.go:447 +0x133

Previous read by goroutine 37:
  sync.raceRead()
      /usr/src/go/src/sync/race.go:37 +0x36
  sync.(*WaitGroup).Add()
      /usr/src/go/src/sync/waitgroup.go:63 +0xc1
  github.com/cockroachdb/cockroach/util.(*Stopper).StartTask()
      /go/src/github.com/cockroachdb/cockroach/util/stopper.go:101 +0x6d
  github.com/cockroachdb/cockroach/storage.(*Range).requestLeaderLease()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:1470 +0x97
  github.com/cockroachdb/cockroach/storage.funcÂ·031()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:1195 +0x4fc
--

tschottdorf commented:
I added a fix for the Stopper because I had a data race.

tbg committed:
wait for elections in kv.createTestDB()

this could help to avoid the following (rare) flakiness:

txn_coord_sender_test.go:82 local sender did not succeed on two attempts: maximum number of attempts exceeded 2

which I've only seen once, but wasn't easy to track down from the logs.

bdarnell commented:
The double increment is not expected to be back; it should be fixed by the fact that we persist the applied index.

bdarnell commented:
FYI the tests that I've seen be most sensitive to tick interval changes are the ones in server/node_test.go.

bdarnell commented on multiraft/multiraft.go:
```diff
@@ -649,6 +649,11 @@ func (s *state) createGroup(groupID uint64) error {
 
+	// Automatically campaign and elect a leader for this group if there's
+	// exactly one known node for this group.
+	if len(cs.Nodes) == 1 && s.MultiRaft.nodeID == NodeID(cs.Nodes[0]) {
```
This may be incorrect if we restarted while adding a second replica: we could have a config change in our log which is committed but unapplied, and we'd need to catch up and apply the log before calling an election. I'm not sure whether this is a real problem though since the second node wouldn't be able to become a leader in this state (membership changes happen one at a time so if the first node crashed while adding the second, there must not be a third). I suppose there is a problem it exists without this change too, since the ticker could fire before the rest of the log has been processed.

Another more serious problem is that Range.InitialState returns a ConfState in which this node is the only member whether we know that to be true or not. We need to populate cs.Nodes in that method only if the range is initialized.

tschottdorf replied:
Good point, I've taken care of populating `cs.Nodes` only for initialized or found ranges - see next commit.

As for conf changes, I just looked at the code and (this is off topic) am a little confused - the conf change is applied immediately when it's appended, but there's no place in which the previous state is stored. It's my understanding that a conf change will have to be undone in some cases (for example if it does not commit and someone without it becomes leader), but I don't see that reflected in multinode/raft.

bdarnell commented:
LGTM

tschottdorf commented:
There's definitely something wrong @bdarnell.
I should be seeing three increments, but with new_value 1, 2 and 3 - correct?
The NodeIDs coming out of this are node1, node3, node5. Sometimes I get node1, node2, node4.

```
14:59 $ docker logs tpc-roachnode1.local 2>&1 | grep 'executed Increment'|grep node-idgen
I0419 12:21:56.119559       9 range.go:915] executed Increment command header:<timestamp:<wall_time:1429446116116274529 logical:0 > cmd_id:<wall_time:1429446116111798546 random:1821423728490400163 > key:"\000node-idgen" end_key:"" user:"root" replica:<node_id:"\001" store_id:"\001" attrs:<> > raft_id:1 read_consistency:CONSISTENT > increment:1 : header:<timestamp:<wall_time:1429446116116274529 logical:0 > > new_value:3
I0419 12:21:56.120470       9 range.go:915] executed Increment command header:<timestamp:<wall_time:1429446116116274529 logical:0 > cmd_id:<wall_time:1429446116111798546 random:1821423728490400163 > key:"\000node-idgen" end_key:"" user:"root" replica:<node_id:"\001" store_id:"\001" attrs:<> > raft_id:1 read_consistency:CONSISTENT > increment:1 : header:<timestamp:<wall_time:1429446116116274529 logical:0 > > new_value:4
I0419 12:21:56.356696       9 range.go:915] executed Increment command header:<timestamp:<wall_time:1429446116355593287 logical:0 > cmd_id:<wall_time:1429446116350990086 random:2166540251145497836 > key:"\000node-idgen" end_key:"" user:"root" replica:<node_id:"\001" store_id:"\001" attrs:<> > raft_id:1 read_consistency:CONSISTENT > increment:1 : header:<timestamp:<wall_time:1429446116355593287 logical:0 > > new_value:5
```

tschottdorf commented:
looks like Raft commands still get executed twice:
```
I0420 13:30:30.014447       6 range.go:915] executed Increment command header:<timestamp:<wall_time:1429536630012030357 logical:0 > cmd_id:<wall_time:1429536630008640719 random:8866370646767159613 > key:"\000node-idgen" end_key:"" user:"root" replica:<node_id:"\001" store_id:"\001" attrs:<> > raft_id:1 read_consistency:CONSISTENT > increment:1 : header:<timestamp:<wall_time:1429536630012030357 logical:0 > > new_value:2
I0420 13:30:30.015015       6 range.go:915] executed Increment command header:<timestamp:<wall_time:1429536630012030357 logical:0 > cmd_id:<wall_time:1429536630008640719 random:8866370646767159613 > key:"\000node-idgen" end_key:"" user:"root" replica:<node_id:"\001" store_id:"\001" attrs:<> > raft_id:1 read_consistency:CONSISTENT > increment:1 : header:<timestamp:<wall_time:1429536630012030357 logical:0 > > new_value:3
```

tschottdorf commented:
I also don't know why the first increment winds up with a new_value of 2. We don't touch that anywhere outside of `server/node.go` and if I log all calls to `r.Increment` on all nodes during the acceptance tests, I only get
```
09:45 $ grep -R "INC " .
./ok1.log:W0420 13:38:02.636672       6 range.go:970] INC "\x00node-idgen" to 2 err=%!s(<nil>)
./ok1.log:W0420 13:38:02.639328       6 range.go:970] INC "\x00store-idgen-2" to 1 err=%!s(<nil>)
./ok1.log:W0420 13:38:03.507254       6 range.go:970] INC "\x00node-idgen" to 3 err=%!s(<nil>)
./ok1.log:W0420 13:38:03.509918       6 range.go:970] INC "\x00store-idgen-3" to 1 err=%!s(<nil>)
```

What am I missing?

tschottdorf commented:
@bdarnell 
> The double increment is not expected to be back; it should be fixed by the fact that we persist the applied index.

But we re-submit all pending commands upon leader change, so that's got to be what's happening. The applied index doesn't help, because it just gets repurposed and just gets a new index. And it bypasses the response cache because we haven't moved up the reproposal mechanism.

Is the above correct?

bdarnell commented:
Ah, right. Re-proposals don't go through the response cache, so that's what's happening. (For some reason I can't reproduce this locally; I reliably get nodes 1 2 and 3).

tschottdorf commented:
It's very sensitive to timing. On my home laptop I get it most of the time.

tbg committed:
feedback by @bdarnell

bdarnell commented:
LGTM

spencerkimball commented on multiraft/heartbeat_test.go:
```diff
@@ -33,11 +31,11 @@ import (
 func processEventsUntil(ch <-chan *interceptMessage, stopper *util.Stopper, f func(*RaftMessageRequest) bool) {
-	if stopper != nil {
-		defer stopper.SetStopped()
+	if stopper == nil {
```
This isn't necessary.

tschottdorf replied:
I've removed it in one of the CLs I'm working on right now.

spencerkimball commented on multiraft/heartbeat_test.go:
```diff
@@ -331,7 +328,6 @@ func TestHeartbeatMultipleGroupsJointLeader(t *testing.T) {
 	}
 	// Keep processing without inspection and shutdown cluster.
-	stopper.AddWorker()
 	go processEventsUntil(transport.Events, stopper, alwaysFalse)
```
Sure this is correct? Why not add a worker here? Maybe use `stopper.RunWorker()`, as you've done elsewhere in this file.

tschottdorf replied:
yeah, this goroutine can just die when the stopper has been invoked (because that means we're in the last line of the test anyways). In the previous code it was waiting to time out and only then let the stopper go ahead.
Could add a worker here again, but it shouldn't make a difference.
